src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
We might be over thinking this a bit.  I can't think of a reason a caller would not want the complete buffer returned to them; and since this abstraction wraps the API call to simplify the behavior and reduce code duplication, it seems strange that we would give the caller a situation where they can make a decision about whether or not the entire data is of interest to them.

However, continuing this conversation, return a stream might make more sense here instead of making a new class.  Fundamentally, we want to read the data back from the transaction, and that is what a stream is for.  The benefit of the stream is that it does not need to read all the data up front and instead the data can be read as needed by the caller.  This solves the problem using an existing Java idiom, and downstream consumers can choose to operate on the data as it is read instead of buffering it all for higher efficiency.  This also solves the potential vulnerability discussed earlier, or at least passes it onto the caller to decide how to handle excessive responses.

Lastly, the stream can be closed to flush the pipe.

The biggest issue we have here once we returned a "delayed" read from the pipe is what you two have touched on around synchronization.  Nobody else can transact against the pipe until the previous transaction is complete.  We can solve that for that caller by adding synchronization around the main transact method in this PR.  If we return a stream, than synchronization in the library does not solve it, and the caller must now perform synchronization of their own to prevent multiple transactions against the same pipe.

We shouldn't be trying to guess, or store state, since that can easily lead to defects or deadlocks.  When in doubt, assume the caller will make a mistake, because they most likely will.

How about we have two methods here.  One that returns all the data, i.e. what we have above, that includes synchronization.  Call it the "easy method".  And a second method that returns a stream, use at your own risk, for high efficiency applications.  The other two methods can be removed at that point.

Thoughts?